### PR TITLE
[DEV-11456] make request chains wait appropriately for sync/async calls

### PR DESCRIPTION
**Pull request by nateshim-indico**
## Summary
Request chains with a wait argument use time.sleep() internally, rather than yielding a delay to the client to sleep synchronously or asynchronously as appropriate. Moved sleeping to client and making all wait calls yield a float or integer denoting sleep time

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] Jira story/task is put into PR title
- [x] Code has been self-reviewed
- [x] Hard-to-understand areas have been commented
- [x] Documentation has been updated
- [x] My changes generate no new warnings
- [x] Unit tests have been added for base functionality and known edge cases
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: indico/client/client.py
Comment (pastap) [inline indico/client/client.py:53]: this might be... a little brittle if the return type is ever a float or int. Have you thought about using something like a "Backoff" class instead? really simple
```
class Backoff:
  def __init__(self, time):
    self.time = time

return Backoff(5)
```
and then you can
```
if isinstance(req, Backoff):
  sleep(Backoff.time)
```

Comment (pastap) [reply indico/client/client.py:53]: or even just leveraging a function to call like
if isinstance(request, Callable): request()...

Comment (pastap) [reply indico/client/client.py:53]: hell yeah, sounds good - for sleeps, could just... 0 debounce or something yeah - great call

File: indico/queries/datasets.py
Comment (nateshim-indico) [inline indico/queries/datasets.py:218]: Is it okay to add here or should this be at bottom of kwarg list

File: indico/client/request.py
Comment (nateshim-indico) [inline indico/client/request.py:93]: Added kwarg `max_wait_time` to all client calls that use `Debouncer` to allow access to change this variable.

File: indico/client/request.py
Comment (pastap) [inline indico/client/request.py:97]: same Union comment here

File: indico/queries/datasets.py
Comment (pastap) [reply indico/queries/datasets.py:218]: good call - bottom of list - just in case we got some folks who blindly pass params

Comment (pastap) [reply indico/queries/datasets.py:218]: and same Union, not Tuple comment

File: indico/queries/export.py
Comment (pastap) [inline indico/queries/export.py:57]: lmao - I'm hoping no one used this otherwise technically a backwards-incompat change. will let @goatrocks make the final call here

File: indico/queries/jobs.py
Comment (pastap) [inline indico/queries/jobs.py:76]: what's the diff? (also Union)

Comment (pastap) [reply indico/queries/jobs.py:76]: also - docstring update?

File: indico/queries/model_groups/model_groups.py
Comment (pastap) [inline indico/queries/model_groups/model_groups.py:31]: docstring update, Union instead of Tuple, etc

File: indico/queries/export.py
Comment (goatrocks) [reply indico/queries/export.py:57]: If we call it out clearly in the release notes, given this isn't something like the submission api... should be ok to fix 🐄  😨

File: indico/queries/jobs.py
Comment (nateshim-indico) [reply indico/queries/jobs.py]: timeout makes call throw `IndicoTimeoutError` after a specified period of time. `max_wait_time` denotes the max wait time per retry call until JobStatus is COMPLETE. 

Updated docstring to hopefully make the distinction clear

File: indico/queries/export.py
Comment (nateshim-indico) [reply indico/queries/export.py:57]: Found this library that support argument deprecation: https://docs.twisted.org/en/stable/api/twisted.python.deprecate.html

Don't know how worth it would be to add it for this

File: indico/queries/datasets.py
Comment (pastap) [inline indico/queries/datasets.py:574]: dumb question - does self.previous pick up the last `yield` in which case it'd mistakenly pick up the debouncer value instead of getdatasetfilestatus?

File: indico/queries/submission.py
Comment (pastap) [inline indico/queries/submission.py:381]: Union

File: indico/queries/datasets.py
Comment (pastap) [inline indico/queries/datasets.py:6]: unneeded Tuple import

File: indico/queries/export.py
Comment (pastap) [reply indico/queries/export.py:57]: or just print a `warnings.warn` if someone uses the old arg?

Comment (goatrocks) [reply indico/queries/export.py:57]: warnings.warn is good -- we have a deprecation library already but unfortunately it doesn't support params (or it didn't when it was added). :(

File: indico/queries/jobs.py
Comment (goatrocks) [reply indico/queries/jobs.py]: `retry_time_interval` or `time_between_retries` would be a clearer name for this variable I think.

File: indico/queries/submission.py
Comment (nateshim-indico) [reply indico/queries/submission.py:381]: smh mb * 5

File: indico/queries/datasets.py
Comment (nateshim-indico) [reply indico/queries/datasets.py:574]: just tested and it does not. I did not think of that tbh so good thing for me to be aware of going forward thanks

File: indico/queries/jobs.py
Comment (nateshim-indico) [inline indico/queries/jobs.py:73]: @goatrocks @pastap  Should we `warnings.warn` before removing?

File: indico/queries/export.py
Comment (nateshim-indico) [reply indico/queries/export.py:60]: added typo argument back and checks for it as discussed

File: indico/queries/jobs.py
Comment (nateshim-indico) [reply indico/queries/jobs.py]: changed `max_wait_time` to `request_interval` to match existing argument in `JobStatus` query

**Review by jacobmanderson — APPROVED** [review]

File: indico/queries/datasets.py
Comment (goatrocks) [inline indico/queries/datasets.py:209]: bless your doc updates :)

File: indico/queries/datasets.py
Comment (goatrocks) [inline indico/queries/datasets.py:550]: 5 seconds? Minutes?

File: indico/queries/export.py
Comment (goatrocks) [reply indico/queries/export.py:60]: Thank you Nate!

**Review by goatrocks — DISMISSED** [review]
lgtm pending @mawelborn 's feedback 👍

**Review by mawelborn — CHANGES_REQUESTED** [review]
Couple issues:

1. [Debouncing](https://underscorejs.org/#debounce) is a specific event-based programming technique that I was surprised to see pop up here. Reading the code, I see it's actually doing something related to backoff.
2. Regardless of the `request_interval` or `Debouncer.max_timeout` specified, the client will always sleep 1 second because that's where the `debouncer` starts, and `debouncer.backoff()` never gets called twice on a single instance.

As someone who creates request chains, but isn't overly familiar with the client codebase, I found this to be a bit confusing. If the desire is to avoid yielding a bare `float` for delays (understandable), my recommendation would be to replace it with a simple `Delay` dataclass to improve readability and fix the issue of not being able to yield a longer delay.

```python
@dataclass
class Delay:
    seconds: float
```

```python
class IndicoClient:
    ...

    def _handle_request_chain(self, chain: RequestChain):
        ...
        elif isinstance(request_or_delay, Delay):
            time.sleep(request_or_delay.seconds):
```

```python
class WaitUntilComplete(RequestChain):
    ...

    def requests(self) -> Iterator[GraphQLRequest | Delay]:
        yield GetSubmission(self.submission_id)
        while self.previous.status != "COMPLETE":
            yield Delay(seconds=8)
            yield GetSubmission(self.submission_id)
```

**Comment by nateshim-indico** [comment]
Debouncer was used for CreateExport call (I don't know the history of why) and so we leveraged it throughout for consistency.

@goatrocks @pastap I'm thinking Debouncer is probably not needed for CreateExport and we can use simple Delay class throughout instead? I don't think it makes sense to have both especially if Debouncer will be used for a single call (CreateExport)

**Review by mawelborn — APPROVED** [review]
Beautiful!

**Review by goatrocks — APPROVED** [review]

**Pull request merged**
